openfisca_uk/calibration/reweighting.py

```diff
@@ -0,0 +1,358 @@
+from pathlib import Path
+import h5py
+from tqdm import tqdm
+from openfisca_uk import Microsimulation
+import pandas as pd
+import numpy as np
+import tensorflow as tf
+import logging
+
+tf.get_logger().setLevel("INFO")
+logging.basicConfig(level=logging.INFO)
+
+sim = Microsimulation(adjust_weights=False)
+parameters = sim.simulation.tax_benefit_system.parameters
+statistics = parameters.calibration
+
+START_YEAR = 2019
+END_YEAR = 2022
+
+FORCE_ALL_METRIC_IMPROVEMENT = False  # Don't save weights unless all (program, year, aggregate/caseload) tests improve. If false, regressions will be logged.
+
+household_population = sim.calc("people", map_to="household").values
+household_region = sim.calc("region").values
+regions = list(pd.Series(household_region).unique())
+household_country = sim.calc("country").values
+countries = list(pd.Series(household_country).unique())
+survey_num_households = len(sim.calc("household_id"))
+
+AGGREGATE_ERROR_PENALTY = 1e-5
+PARTICIPATION_ERROR_PENALTY = 1e-1  # Person-deviations are 10,000 more loss-heavy than spending deviations per pound
+MODIFICATION_PENALTY = 1e-11
+AGE_DISTRIBUTION_PENALTY = 1e-1
+POPULATION_ERROR_PENALTY = 5
+
+
+def loss(
+    weight_modification: np.array, include_modification_penalty: float = True
+) -> float:
+    l = 0
+    for year in range(START_YEAR, END_YEAR + 1):
+        instant_str = f"{year}-01-01"
+        statistic_set = statistics(instant_str)
+        variables = list(statistic_set.aggregate._children)
+        weights = sim.calc("household_weight", period=year).values
+        modified_weights = tf.nn.relu(
+            weight_modification[year - START_YEAR] + weights
+        )
+        for variable in variables:
+            if variable in statistic_set.aggregate._children:
+                values = sim.calc(variable, period=year).values
+                entity = sim.simulation.tax_benefit_system.variables[
+                    variable
+                ].entity.key
+                household_totals = sim.map_to(values, entity, "household")
+                household_participants = sim.map_to(
+                    (values > 0) * 1, entity, "household"
+                )
+                # Calculate aggregate error
+                agg = tf.reduce_sum(modified_weights * household_totals)
+                target = getattr(statistic_set.aggregate, variable)
+                l += (
+                    AGGREGATE_ERROR_PENALTY
+                    * target
+                    * ((agg / target) - 1) ** 2
```

Comment:
Could you make a function for this e.g. `squared_relative_deviation(value, target)`? Since it's used below too. Might be room for modularizing other stuff too, but this would at least be a bit more descriptive.